### DISCUSSION TITLE
[Solved] FastDDS Discovery Server Config Pi5 ROS 2 in Docker

### DISCUSSION THREAD

**Discussion by slowrunner** [iRobotEducation/create3_docs#549] [Questions] [answered]
# [Solved] [See Here](https://github.com/iRobotEducation/create3_docs/discussions/549#discussioncomment-8510583)

### How are you connecting to your Create 3?

Wi-Fi and USB Ethernet (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Ubuntu 22.04 in Docker over Raspberry Pi 5 PiOS Bookworm

### Which version of ROS 2 is installed on your computer?

Humble

### Which firmware version is installed on your robot?

H.2.4

### Which RMW is your robot running?

FastRTPS

### Is the robot connected to a network? If so, what is the network type?

Yes: 2.4GHz Home WiFi 

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

Yes

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

# [SOLVED - [See Reply](https://github.com/iRobotEducation/create3_docs/discussions/549#discussioncomment-8510583)]

I don't think the Create3 is "seeing" my fast dds discovery server - probably needs some special network routing.

Simple Discovery Configuration Works:
- The Raspberry Pi5 runs PiOS and connects to my 5GHz WiFi as 10.0.0.129
- ROS is running inside a Docker Container with host network mapped
- wali_node.py publishes and subscribes topics over 5GHz WiFi
- 5GHz WiFi routes to/from 2.4GHz WiFi
- Create3 publishes/subscribes topics over 2.4GHz WiFi

Fast DDS Discovery Server Configuration Not Working:
- The Raspberry Pi5 runs PiOS and connects to my 5GHz WiFi as 10.0.0.129
- The Raspberry Pi5 has usb0 network address 192.168.186.3 and can see Create3 at 192.168.186.2
- ROS is running inside a Docker Container with "host networking" mapped
- Discovery Server running inside Docker Container with -l 192.168.186.3 -p 11811
- 5GHz WiFi routes to/from 2.4GHz WiFi

- Create3 publishes/subscribes topics with fastdds discovery service enabled with 192.168.186.3:11811

Have the feeling the packets from create3 don't see the discovery server inside the Docker ROS container

**Comment by slowrunner** [comment, marked as answer]
# FastDDS Discovery Server Config to support ROS 2 in Raspberry Pi 5 Docker Container For Create3

![DiscoveryServerNTPconfig](https://github.com/iRobotEducation/create3_docs/assets/8950202/1fa3134f-e4c2-4e70-b553-969a775ae4bf)


Concept:  
- Discovery Server is running on the Pi5 inside Docker container 
  (because no supported ROS 2 Humble for  Ubuntu 23.04)
- Discovery Server has three ports
  - one for Create3, 
  - one for Pi5 local nodes and CLI sessions, 
  - one for WiFi connected clients such as Desktop Rviz2 or CLI terminal sessions.  
  
Each environment has a configuration/file telling nodes and sessions where in the "network" to find the discovery server. 
- ```export ROS_DISCOVERY_SERVER=x.x.x.x:11811``` tells nodes where to find the discovery server
- ```export FASTRTPS_DEFAULT_PROFILES_FILE=/home/pi/wali_pi5/configs/super_client_configuration_file.xml``` tell CLI command line ROS commands where to find the discovery server??  

I don't actually understand these two exports, UDP, TCP/IP, and ROS 2 messaging.  Just saying.

Step-by-step coming after I recover from bashing my head over this, the ntp setup, and the Pi5 usb0 port setup.

**Reply by alsora** [reply]
This is great!
Please consider submitting a pull request to the create 3 docs with the complete step-by-step, it would be greatly appreciated!

**Reply by slowrunner** [reply]
# Step by Step (PI5 - ROS 2 HUMBLE FastDDS IN DOCKER )

Starting point: Create3, Pi5, and Desktop all running ROS 2 Humble FastDDS "Simple Discovery" - Full visibility on all three.

6)  Setup Discovery Server to start in WaLiPi5 Dockered ROS:
- Pi5: added configs/super_client_configuration_file.xml with IP of 127.0.0.1:11811

```
<?xml version="1.0" encoding="UTF-8" ?>
<dds>
    <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
        <participant profile_name="super_client_profile" is_default_profile="true">
            <rtps>
                <builtin>
                    <discovery_config>
                        <discoveryProtocol>SUPER_CLIENT</discoveryProtocol>
                        <discoveryServersList>
                            <RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
                                <metatrafficUnicastLocatorList>
                                    <locator>
                                        <udpv4>
                                            <address>127.0.0.1</address>
                                            <port>11811</port>
                                        </udpv4>
                                    </locator>
                                </metatrafficUnicastLocatorList>
                            </RemoteServer>
                        </discoveryServersList>
                    </discovery_config>
                </builtin>
            </rtps>
        </participant>
    </profiles>
</dds>

```
- Pi5: added to end of ~/.bashrc:

```
export ROS_DISCOVERY_SERVER="127.0.0.1:11811"
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/pi/wali_pi5/configs/super_client_configuration_file.xml
```

- Pi5: added to start_wali.sh to run in background with a 5s sleep  (replaced pi5_wifi_IP with actual IPv4 x.x.x.x)

```
# echo -e "\n*** Start FastDDS Discovery Server"
echo '*** fastdds discovery -i 0 -l 192.168.186.3 -p 11811 -l pi5_wifi_IP -p 11811 -l 127.0.0.1 -p 11811 &'
fasts discovery -i 0 -l 192.168.186.3 -p 11811 -l pi5_wifi_IP -p 11811 -l 127.0.0.1 -p 11811 &
export ROS_DISCOVERY_SERVER="127.0.0.1:11811"
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/pi/wali_pi5/configs/super_client_configuration_file.xml
```

7)  Create3: Enable Fast DDS discovery server in Create3 Application Configuration
    - check Enable Fast DDS discovery server? box
    - Enter 192.168.186.3:11811  (looks like it is there but need to enter it)
    - Click Save
    - Click Restart Application

8)  Configure desktop ROS CLI and any nodes to be "SuperClients" of discovery server:

- Desktop: Create file wali_desk/config/super_client_configuration_file.xml: (replaced PI5_WIFI_IP with actual IPv4 x.x.x.x)

```
<?xml version="1.0" encoding="UTF-8" ?>
<dds>
    <profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
        <participant profile_name="super_client_profile" is_default_profile="true">
            <rtps>
                <builtin>
                    <discovery_config>
                        <discoveryProtocol>SUPER_CLIENT</discoveryProtocol>
                        <discoveryServersList>
                            <RemoteServer prefix="44.53.00.5f.45.50.52.4f.53.49.4d.41">
                                <metatrafficUnicastLocatorList>
                                    <locator>
                                        <udpv4>
                                            <address>PI5_WIFI_IP</address>
                                            <port>11811</port>
                                        </udpv4>
                                    </locator>
                                </metatrafficUnicastLocatorList>
                            </RemoteServer>
                        </discoveryServersList>
                    </discovery_config>
                </builtin>
            </rtps>
        </participant>
    </profiles>
</dds>


```
- Desktop: add to ~/.bashrc:

```
export FASTRTPS_DEFAULT_PROFILES_FILE=/home/ubuntu/wali_desk/config/super_client_configuration_file.xml
export ROS_DISCOVERY_SERVER=10.0.0.219:11811

```

8)  Pi5: Confirm Create3 topics visible on WaliPi5 (/dock_status, /odom, /ir_intensity)
    - Confirm immediately
    - Confirm after doing a warm boot of Pi5
    - Confirm after doing a storage/ship shutdown, boot of Create3 and Pi5

9)  Pi5: Confirm WaliPi5 topics seen by Create3 (publisth /cmd_vel, publish /dock)

10) Desktop: Confirm Create3 topics reach desktop (/battery_state, /dock_status) and rviz (/odom)

11) Desktop: Confirm WaliPi5 topics reach rviz (/mapUpdate, /oak/color/image_raw,  /scan)

12) Desktop: Confirm Desktop topics reach Create3 (pub /dock)
